packages/snjs/lib/services/settings_service/SettingsRepo.ts

```diff
@@ -0,0 +1,100 @@
+import { SettingName, SettingPayload } from '@standardnotes/settings';
+import * as messages from '../api/messages';
+import { SettingsProvider } from './SettingsProvider';
+import {
+  DeleteSettingResponse,
+  GetSettingResponse,
+  ListSettingsResponse,
+  UpdateSettingResponse,
+  User,
+} from '../api/responses';
+import { UuidString } from '@Lib/types';
+
+interface SettingsAPI {
+  listSettings(userUuid: UuidString): Promise<ListSettingsResponse>;
+
+  updateSetting(
+    userUuid: UuidString,
+    settingName: string,
+    settingValue: string
+  ): Promise<UpdateSettingResponse>;
+
+  getSetting(
+    userUuid: UuidString,
+    settingName: string
+  ): Promise<GetSettingResponse>;
+
+  deleteSetting(
+    userUuid: UuidString,
+    settingName: string
+  ): Promise<DeleteSettingResponse>;
+}
+
+export class SettingsRepo implements SettingsProvider {
+  constructor(
+    private readonly api: SettingsAPI,
+    private readonly getUser: () => User | null
+  ) {}
+
+  canModify() {
+    return this.getUser() != null;
+  }
+
+  private get userUuid() {
+    const user = this.getUser();
+    if (user == null || user.uuid == null) {
+      throw new Error(messages.API_MESSAGE_INVALID_SESSION);
+    }
+    return user.uuid;
+  }
+
+  async listSettings() {
+    const { error, settings } = await this.api.listSettings(this.userUuid);
+    if (error != null) throw new Error(error.message);
+    if (settings == null) return {};
+
+    const settingVals = Object.values<string>(SettingName);
+
+    const res: Partial<SettingPayload> = {};
+    for (const s of settings) {
+      if (settingVals.includes(s.name) && s.value != null) {
+        const name = s.name as SettingName;
+        if (s.value != null) {
+          (res as any)[name] = JSON.parse(s.value);
+        }
+      }
+    }
+    return res;
+  }
+
+  async getSetting<Key extends keyof SettingPayload>(name: Key) {
+    const { error, setting } = await this.api.getSetting(this.userUuid, name);
+
+    if (error != null) throw new Error(error.message);
+    if (setting?.value == null) return null;
+    return JSON.parse(setting.value);
+  }
+
+  async updateSetting<Key extends keyof SettingPayload>(
+    name: Key,
+    payload: SettingPayload[Key]
+  ): Promise<SettingPayload[Key] | null> {
+    const value = JSON.stringify(payload);
```

Comment:
I think it would be better for only the final settings repository to be JSON stringified/parsed, rather than stringifying and parsing every value. This is how storage service operates and it makes for a less confusing API.

How are settings persisted to disk? You may not even need to think about JSON if you are passing the final repository to StorageService.

Reply:
Yep, you're right! It's fixed now.

The settings aren't persisted yet in this PR. Wanted to get the test going with the direct access first.